### PR TITLE
Readonly wrapper

### PR DESCRIPTION
**Pull request by anton-larsen-persequor**
- This is still experimental

### CODE REVIEW

File: ran-core/src/main/java/io/ran/Clazz.java [hunk 1]
```diff
@@ -37,6 +32,9 @@ public static Clazz of(Type type) {
 			return Clazz.ofClazzes((Class)((ParameterizedType) type).getRawType(), genericClasses);
 		} else if (type instanceof Class) {
 			return Clazz.of((Class)type);
+		} else if (type instanceof TypeVariable || type instanceof WildcardType) {
+//			TypeVariable typeVariable = (TypeVariable) type;
+			return Clazz.of(Object.class);
```
Comment (emanuel-persequor) [inline ran-core/src/main/java/io/ran/Clazz.java:37]: I guess this is just a general fix to ran?

File: ran-core/src/main/java/io/ran/Clazz.java [hunk 2]
```diff
@@ -125,8 +123,11 @@ public String getDescriptor() {
 		if (isPrimitive()) {
 			return Primitives.get(clazz).getDescriptor();
 		}
-		if (clazz == byte[].class) {
-			return "[B";
+		if (clazz.isArray()) {
```
Comment (emanuel-persequor) [inline ran-core/src/main/java/io/ran/Clazz.java:126]: I guess this is just a general improvement to ran?

File: ran-core/src/main/java/io/ran/Clazz.java [hunk 3]
```diff
@@ -135,8 +136,12 @@ public String getSignature() {
 		if (isPrimitive()) {
 			return Primitives.get(clazz).getDescriptor();
 		}
-		if (clazz == byte[].class) {
-			return "[B";
+
+		if (clazz.isArray()) {
```
Comment (emanuel-persequor) [inline ran-core/src/main/java/io/ran/Clazz.java:140]: I guess this is just a general improvement to ran?

File: ran-core/src/main/java/io/ran/ReadonlyWrapper.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ReadonlyWrapper {
+	private static Map<Class, Class> wrapped = new HashMap<>();
```
Comment (emanuel-persequor) [inline ran-core/src/main/java/io/ran/ReadonlyWrapper.java:16]: static but not thread-safe/synchronized.

File: ran-core/src/main/java/io/ran/ReadonlyWrapper.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+			throw new RuntimeException(e);
+		}
+	}
+
+	private static  <T, W extends T> Class<W> internalWrap(Class<W> wc) {
+		Class t = (List.class.isAssignableFrom(wc)) ? ArrayList.class : wc;
+
+		return wrapped.computeIfAbsent(wc, c -> {
+			try {
+				Path path = Paths.get("/tmp/" + wc.getSimpleName() + "Readonly.class");
```
Comment (emanuel-persequor) [inline ran-core/src/main/java/io/ran/ReadonlyWrapper.java:37]: Linux specific? https://mkyong.com/java/how-to-get-the-temporary-file-path-in-java/ (or similar, maybe even wrap "getTempDir()" in some kind of injected little helper.

Comment (MikePSQR) [reply ran-core/src/main/java/io/ran/ReadonlyWrapper.java:37]: that is a debug thing, no?

**Pull request closed without merging**